### ISSUE TITLE
Add OBO middleware [Feature request]

### ISSUE THREAD

**Issue by lassebass** [Intility/fastapi-azure-auth#100]
**Describe the feature you'd like**
It would be nice to have the framework allow more than one API scope. The case where this is needed, is if you have multiple business applications that need access to the same API. (For example like how both Teams and Outlook has access to your calendar). Features like this is supported in .NET frameworks by making a list of valid isuers and audiences, instead of enforcing just one (see [this StackOverflow example](https://stackoverflow.com/questions/46990509/how-to-set-multiple-audiences-in-asp-net-core-2-0-addjwtbearer-middleware) for how the .NET `AddJwtBearer` middleware works)

The "correct way" of dealing with these cases is to add middleware implementing the [OBO](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) ("On-Behalf-Of") flow. Usually this is handled by the client, but for third party applications and/or plugins, we cannot initiate OBO client-side.

**Additional context**
If I supply an access token with an audience that I've added to the my application's `knownClientApplications` list, the token validation should pass.

**Comment by h3rmanj** [comment]
I'm not sure I understand this question.

The audience of the access token is the application's `clientId`, not the consumers `clientId`. You shouldn't have to add an application to it's own `knownClientApplications` list. The consumers `clientId` is not validated in this package (and isn't in .NET by default either), so who can create access tokens to your application is dependent of what the issuer allows.

Application that isn't multi-tenant but still has separate app registrations in different tenants something that could be implemented though, but I'm not sure that's what you're asking.

**Comment by lassebass** [comment]
We have multiple App Registrations set up for authentication over many different applications in our Azure portfolio. The feature I'm looking for, is the option to allow _fastapi_azure_auth_ to accept tokens issued to any one of these app registrations, not just one specific one.

That way, as long as you are logged into any one of our app registrations, you can access the Web API.

With the current setup, _fastapi_azure_auth_ only allows **one** `app_client_id` in a `SingleTenantAzureAuthenticationCodeBearer`. A consequence of this is that the API will only accept tokens that are logged into that single app registration.

I.e. expand this

```python
azure_scheme = SingleTenantAzureAuthorizationCodeBearer(
    app_client_id="00001111-2222-3333-4444-555566667777",
    tenant_id="88889999-aaaa-bbbb-cccc-ddddeeeeffff",
    scopes={
        "api://00001111-2222-3333-4444-5555666667777/.default"
    }
```

to something like


```python
azure_scheme = SingleTenantAzureAuthorizationCodeBearer(
    app_client_ids=[
        "00001111-2222-3333-4444-555566667777",
        "00001111-2222-3333-4444-555566667778",
        ...
    ],
    tenant_id="88889999-aaaa-bbbb-cccc-ddddeeeeffff",
    scopes={
        "api://00001111-2222-3333-4444-5555666667777/.default",
        "api://00001111-2222-3333-4444-5555666667778/.default",
        ...
    }
```

**Comment by h3rmanj** [comment]
I'm still unsure as to why you would want to accept access tokens issued to another application. If the consumers have been granted the scopes to the API in Azure, they should have no problems requesting tokens to the API without any additional configuration in the API.

**Comment by JonasKs** [comment]
Sorry, I’ve been away so didn’t have time to look into this. 

I’m also struggling to understand the use case here.  

> The feature I'm looking for, is the option to allow fastapi_azure_auth to accept tokens issued to any one of these app registrations, not just one specific one.

Can you give me some specific examples of a setup here? Or Microsoft documentation?   
Seems to me that your application should just [fetch multiple tokens](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-user-gets-consent-for-multiple-resources).

**Comment by lassebass** [comment]
TL;DR: fetching multiple tokens is not an option.

Use case: users are logged into a [Grafana](https://azure.microsoft.com/en-us/services/managed-grafana) instance.

Since this is an out-of-the-box Azure product, we cannot directly control how this app handles user login sessions, i.e. cannot fetch multiple tokens. OAuth setup in Grafana connects to one specific app registration (the secondary app), which is strictly different to our FastAPI instance (primary app).

The current state of the fastapi-azure-auth library makes it great as a standalone product, but makes it somewhat lacking when trying to plug in with pre-existing solutions. For now we'll have to make do with some workarounds (which, if you're interested, is to create two **azure_acheme**s, and split the FastAPI instance in two)

**Comment by fossmark** [comment]
The reason why this could be a nice feature is that if you want to integrate with a third party software which you do not develop and have control over .

One example could be the JSON Plugin to Grafana which makes it possible to execute requests against any arbitrary backend and this could forward the JWT token the browser already has.

I agree that it would be more correct to use the On-Behalf-Of flow and fetch a new token which has the expected audience - but as a practical and pragmatic solution it would be nice to have a feature that allows use to specify multiple audiences.

In the possible to specify multiple audience in for example Dotnet Core using the TokenValidationParameters class so I do not think this is a very specific and limited use case.

https://docs.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.tokens.tokenvalidationparameters.validaudiences?view=azure-dotnet#microsoft-identitymodel-tokens-tokenvalidationparameters-validaudiences

**Comment by JonasKs** [comment]
Who will be the issuer of these tokens? Your tenant or a generic Microsoft tenant (making it multi-tenant)? As far as I can see, the only parameter that we need to support a list for is the [audience](https://github.com/Intility/fastapi-azure-auth/blob/main/fastapi_azure_auth/auth.py#L187). 

Since we don't actually manually pass in `audience`, I think this could be an `audience: list[str] | None = None` parameter. This way we don't have to change any documentation or have any breaking changes, but one could manually pass in `audience` if one wants to.   
In other words, [this](https://github.com/Intility/fastapi-azure-auth/blob/main/fastapi_azure_auth/auth.py#L187) should be `audience or client_id if token_version == 2 else f'api://{client_id}'`... Unfortunately, I checked `python-jose`, and it does not support a list of audiences. In other words, we'd have to do audience verification manually, or (preferabily) try to implement this in `python-jose`.

**Comment by JonasKs** [comment]
https://github.com/mpdavis/python-jose/issues/302

**Comment by JonasKs** [comment]
Closing. PRs welcome, but I do not intend to implement this without `python-jose` request getting approved.
